src/imitation/scripts/multi_train_adversarial.py

```diff
@@ -0,0 +1,149 @@
+
+  Returns:
+    `row` (same as argument), and `result` (the return value of the run).
+  """
+  run = train_ex.run
+  stdout_path = osp.join(log_dir, "stdout")
+  stderr_path = osp.join(log_dir, "stderr")
+  run = multi_util.redirect_to_files(run, stdout_path, stderr_path)
+
+  named_configs = [row['env_config']] + list(extra_named_configs)
+  config_updates = {
+    "log_dir": log_dir,
+    "n_gen_steps_per_epoch": int(row['n_gen_steps_per_epoch']),
+    "init_trainer_kwargs.n_expert_demos": int(row['n_expert_demos']),
+    "seed": int(row['phase3_seed']),
+  }
+  config_updates.update(**extra_config_updates)
+
+  result = run(
+    named_configs=named_configs, config_updates=config_updates).result
+  return row, result
+
+
+@multi_train_ex.main
+def benchmark_adversarial_from_csv(
+  log_dir: str,
+  n_workers: int,
+  csv_config_path: str = "experiments/gail_benchmark_config.csv",
+  seeds: Iterable[int] = range(3),
+  extra_named_configs: Optional[Iterable[str]] = None,
+  extra_config_updates: Optional[dict] = None,
+  parallel: bool = True,
+) -> str:
+  """Start several runs of `scripts.train_adversarial` from a CSV file.
+
+  Results are written as a CSV to `f"{log_dir}/results.csv"`.
+  The columns of the results file are the columns of the CSV config file,
+  followed by "phase3_seed", and then the keys of the
+  `train_adversarial.run(...).result`.
+
+  Params:
+    log_dir: Main logging directory.
+    n_workers: The number of jobs that are processed simultaneously.
+    csv_config_path: Path to CSV configuration. Columns should include
+      'env_config' (str), n_gen_steps_per_epoch (int), 'n_expert_demos' (int),
+      and 'phase3_seed' (int).
+    seeds: Every row in `csv_config_path` is run using each of these seeds.
+      The seed associated with each row in the output CSV,
+      `f"{log_dir}/results.csv"`, is saved in a column called "phase3_seed".
+    extra_named_configs: Extra named configs to apply to every call of
+      `train_adversarial.run(...)`.
+    extra_config_updates: Extra config updates to apply to every call of
+      `train_adversarial.run(...)`.
+    parallel: If True, then uses `n_workers` processes. If False, then use
+      n_workers threads (mostly useful for debugging purposes).
+
+  Returns:
+    Path to the output CSV file, `f"{log_dir}/results.csv"`.
+  """
+  if extra_named_configs is None:
+    extra_named_configs = []
+  if extra_config_updates is None:
+    extra_config_updates = {}
+  if parallel:
+    mp = multiprocessing
+  else:
+    mp = multiprocessing.dummy
+
+  # Construct job args for `Pool.imap_unnordered`.
+  # There is one job for every combination of seed and CSV row.
+  job_args = []  # type: List[tuple]
+  with open(csv_config_path, newline='') as csv_file:
+    for row in csv.DictReader(csv_file):
+      for seed in seeds:
+        new_row = row.copy()  # type: OrderedDict
+        new_row["phase3_seed"] = str(seed)
```

Comment:
Why is this phase3? Isn't this after `expert_demos` which was phase4?

Reply:
Thanks for clarifying, this diagram is really useful. Might we want to use `multi_expert_demos` for phase 1 as well as phase 4 though?

Reply:
yes, this would be nice. Phase 1 is currently performed by `experiments/train_experts.sh`. When I made this PR I intended to punt this issue till later (another thing that would have been good to mention in the PR summary)

Reply:
For completion I'll write here that Phase 2 is roughly done by #87